### DISCUSSION TITLE
[Idea] astro/icons

### DISCUSSION THREAD

**Discussion by FredKSchott** [withastro/roadmap#107] [Proposal]
Hey everyone! This was a part of my exploration into assets that I didn't want to lose, so I took some quick notes using the RFC template as a guide.

I think this is a cool feature, but I would love to get more feedback on it before I clean this up and submit an official RFC. **Is this something that people would like to see?**

If you are feeling adventurous you can check this out and use it locally via this branch: https://github.com/withastro/astro/tree/astro-icons

# Summary

A built-in icon set for Astro users.

# Example

```astro
---
import {Icon} from 'astro/components';
import {IconRocket, IconCookie, IconSearch} from 'astro/icons';
---
<div>
    <Icon use={IconRocket} />
    <Icon use={IconCookie} size={32} />
    <Icon use={IconSearch} size={64} />
</div>
```

# Motivation

- Icons are a common use-case for web developers. 
- Finding a good icon set can be difficult that involves a lot of googling. How do I choose the best one? 
- Some external icon sets are poorly implemented and cause an overloading of data. For example, some are shipped as React components or require svgo to convert SVGs into React. Others use webfonts instead of SVGs, which is more expensive.

What if Astro came with a single, solid icon set built-in? This is slightly different from the astro-icon exploration that @natemoo-re has been doing, and shouldn't be seen as conflicting. This does not attempt to help you load your own SVGs, or help you use your favorite icon set with Astro. This is a single, opinionated icon set available to all Astro users.

# Detailed design

## The `astro/icons` Import

This RFC proposes an "astro/icons" entrypoint that is generated from the Google Material Design icons. It essentially looks like this:

```
export const IconCookie = "<path d=\"M21.95 10.99c-1.79-.03-3.7-1...";
export const IconCopyright = "<path d=\"M11.88 9.14c1.28.06 1.61 ...";
export const IconCoronavirus = "<path d=\"M21.25 10.5c-.41 0-.75 ...";
// ...
export const IconAirbnbLogo = "<path d=\"M224 373.12c-25.24-31.67...";
export const IconAlgoliaLogo = "<path d=\"M229.3 182.6c-49.3 0-89...";
export const IconAlipayLogo = "<path d=\"M377.74 32H70.26C31.41 3...";
```

It currently contains:
  - 3382 different icons, generated from Google's material icons collection
  - 460 logos, generated from FontAwesome's brands collection

And it supports:
  - Intellisense, so that VSCode will auto-complete your imports.
  - TypeScript, so that your typechecker will fail if you use a bad icon.
  - Treeshaking, you will only pay the cost for the exact icons that you use.

## The `<Icon>` Component

This RFC proposes a basic `<Icon>` component that takes the SVG contents that are returned from the `astro/icons` import and wraps them in an `<svg>` element.

```
---
// Example Implementation
export interface Props {
	use: string;
	size?: number;
}
let { use, size = 32 } = Astro.props;
---
<svg height="${size}" width="${size}" set:html={use} />
```

# Drawbacks

The biggest drawback is the long-term commitment to picking a single icon set. If we were to change this later, it would be expensive (but not impossible) to change sets.

# Alternatives - Approach

- We could just not do this!

# Alternatives - Other Icon Sets

Several icon sets are available to bring into Astro. Here is the research that I did on some other options:

#### Notable Contenders:
  - Heroicons: Well-designed, Maintained by Tailwind but not under active dev, 230 icons is a little small.
  - Fontawesome: Well maintained and actively developed, well-stocked. Biggest reason against is that the style is really well-recognized, and would be hard for us to feel like this was "Astro Icons" vs. "Astro comes with Fontawesome built-in". Maybe some would see that as a perk, but Google's Material Icon is equally well maintained and well stocked, but with a much more distinctive style.

#### Others Considered:
- Feather Icons - not under active dev, 238 icons is a little small
- Bootstrap - 1600 icons is well-sized, but it didn't have some basic icons that I searched for ("flame", "rocket").

# Adoption strategy

- Not a breaking change.

# Unresolved questions

None yet.

**Comment by zadeviggers** [comment]
I don't think Astro should single out and use a single icon set. I would much prefer something similar to `astro-icon` where (almost) any icon set can be used. That would enable a lot more creative freedom.

**Reply by retronav** [reply]
This. Also, though the idea of providing icons is nice, I would prefer having them in a separate package, because those would be practically unused and contribute to bloat unless one uses them.

**Reply by FredKSchott** [reply]
See response here!
https://github.com/withastro/rfcs/discussions/107#discussioncomment-2175304

**Comment by natemoo-re** [comment]
I am biased here, but I much prefer the approach of `astro-icon` and wouldn't like to see something like `astro/icons` that is less flexible and powerful. For small personal sites, I'm sure this built-in set would be a minor convenience. For most professional sites, folks will inevitably reach for their own designs, meaning `astro/icons` would just be bloat.

I also _really_ want to avoid the mistake every other tool has made where you need to publish your SVG icons in a special "Astro" format in order for them to work. Even though this is just `export const IconName = "stringified-path";`, it is still another format that the author needs to generate from the original `.svg` files. Astro _should_ just be able to load `.svg` files.

**Reply by ghost** [reply]
Agreed on this! I will poke a small hole in `astro-icon` as well though: I think my biggest complaint about Astro UI primitives is their `.astro` restriction. For instance, I'd naturally want to reach for an icon library when building a hamburger menu. But if I build that hamburger trigger in a Svelte component... welp, guess I need a Svelte icon library now 😕 

I think @jonathantneal's work on import assertions could lead to a more sustainable solution. For instance, a way to import `.svg` files _as_ a given component syntax.

**Reply by FredKSchott** [reply]
> I also really want to avoid the mistake every other tool has made where you need to publish your SVG icons in a special "Astro" format in order for them to work.

+1 for supporting both `<svg>...` and `<path>...`! This implementation uses only the path to reduce size and also make it easier to customize the svg wrapper (height, width, etc). But whatever "component" uses these svgs should be flexible enough to also support a full `<svg>` element.

**Reply by FredKSchott** [reply]
>  I much prefer the approach of astro-icon

I hope this idea doesn't come across as "anti- astro-icon"! Strong +1 on having a great story on working with any other icon pack. The idea here was to have something built-in to Astro that is a great set of icons for someone who doesn't yet have a favorite (see motivation). I personally just want great icons, and the promise of "astro-icons" letting me use any public icon pack doesn't really help me since I don't know which are good and which are bad. If this is a different problem for a different kind of user, then I think that the two can live together in harmony.

1 improvement could be to combine these two ideas into one proposal/solution: have a great built-in, zero-thought icon pack AND a great, built-in way to use your favorite icon pack. I can see how calling this proposal `astro/icons` doesn't leave much room for anything else! @natemoo-re @Holben888 if we expanded this to also incorporate the "any icon pack" ideas of astro-icon, would that address your concern?

**Reply by zadeviggers** [reply]
The Icon component (assuming it's the same as `astro-icon`) could just have an optional pack property that's set to something like font awsome by default.

**Comment by naiyerasif** [comment]
It would be fantastic if we can introduce an `sprite` mode which can simply embed all those icons as `symbol`s somewhere in the generated document or even export them as an external asset. Then we can simply `use` them through the `symbol`'s id. This would help avoid copying the html, and might be faster if the icon set is massive.

```astro
---
export interface SpriteProps {
	path?: string;
	size?: number;
}
let { path, size = 32 } = Astro.props as SpriteProps;
---
<svg height="${size}" width="${size}">
	<use href={path} />
</svg>
```

**Reply by FredKSchott** [reply]
Reminder that only the used icons are included in your build, so the size of the overall icon set shouldn't impact final HTML size.

+1 for this idea if users want it, astro-icon currently supports this. I tried to focus this discussion on "how we access built-in icons" and avoid the implementation of getting the SVG on the page for now.

**Comment by FredKSchott** [comment]
Thanks everyone for the feedback! This is something I personally would love to explore more but I don't think aligns with our pre-v1.0 goals. I'll explore as a community component for now, and we can revisit post v1.0 if people love it.
